src/extensions/mvs/helpers/indexing.ts

```diff
@@ -0,0 +1,130 @@
+ * @author Adam Midlik <midlik@gmail.com>
+ */
+
+import { Column } from '../../../mol-data/db';
+import { SortedArray } from '../../../mol-data/int';
+import { ChainIndex, ElementIndex, Model, ResidueIndex } from '../../../mol-model/structure';
+import { Mapping, MultiMap, NumberMap, filterInPlace, range, sortIfNeeded } from './utils';
+
+
+/** Auxiliary data structure for efficiently finding chains/residues/atoms in a model by their properties */
+export interface IndicesAndSortings {
+    chainsByLabelEntityId: Mapping<string, readonly ChainIndex[]>,
+    chainsByLabelAsymId: Mapping<string, readonly ChainIndex[]>,
+    chainsByAuthAsymId: Mapping<string, readonly ChainIndex[]>,
+    residuesSortedByLabelSeqId: Mapping<ChainIndex, Sorting<ResidueIndex, number>>,
+    residuesSortedByAuthSeqId: Mapping<ChainIndex, Sorting<ResidueIndex, number>>,
+    residuesByInsCode: Mapping<ChainIndex, Mapping<string, readonly ResidueIndex[]>>,
+    atomsById: Mapping<number, ElementIndex>,
+    atomsByIndex: Mapping<number, ElementIndex>,
+}
+
+export const IndicesAndSortings = {
+    /** Get `IndicesAndSortings` for a model (use a cached value or create if not available yet) */
+    get(model: Model): IndicesAndSortings {
+        return model._dynamicPropertyData['indices-and-sortings'] ??= IndicesAndSortings.create(model);
+    },
+
+    /** Create `IndicesAndSortings` for a model */
+    create(model: Model): IndicesAndSortings {
+        const h = model.atomicHierarchy;
+        const nAtoms = h.atoms._rowCount;
+        const nResidues = h.residues._rowCount;
+        const nChains = h.chains._rowCount;
+        const { label_entity_id, label_asym_id, auth_asym_id } = h.chains;
+        const { label_seq_id, auth_seq_id, pdbx_PDB_ins_code } = h.residues;
+        const { Present } = Column.ValueKind;
+
+        const chainsByLabelEntityId = new MultiMap<string, ChainIndex>();
+        const chainsByLabelAsymId = new MultiMap<string, ChainIndex>();
+        const chainsByAuthAsymId = new MultiMap<string, ChainIndex>();
+        const residuesSortedByLabelSeqId = new Map<ChainIndex, Sorting<ResidueIndex, number>>();
+        const residuesSortedByAuthSeqId = new Map<ChainIndex, Sorting<ResidueIndex, number>>();
+        const residuesByInsCode = new Map<ChainIndex, MultiMap<string, ResidueIndex>>();
+        const atomsById = new NumberMap<number, ElementIndex>(nAtoms + 1);
+        const atomsByIndex = new NumberMap<number, ElementIndex>(nAtoms);
+
+        for (let iChain = 0 as ChainIndex; iChain < nChains; iChain++) {
+            chainsByLabelEntityId.add(label_entity_id.value(iChain), iChain);
+            chainsByLabelAsymId.add(label_asym_id.value(iChain), iChain);
+            chainsByAuthAsymId.add(auth_asym_id.value(iChain), iChain);
+
+            const iResFrom = h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain]];
+            const iResTo = h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain + 1]] ?? nResidues;
```

Comment:
I think this is a bug. If you want exclusive "end residue" this should be 

```ts
// atomEnd = h.chainAtomSegments.offsets[iChain + 1] is the 1st atom of the "next chain"
// lastResidue = h.residueAtomSegments.index[atomEnd - 1]
// residueUpperBound = lastResidue + 1
//   =>
h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain + 1] - 1] + 1
```

The joys of indexing :)

Reply:
I'm taking "the first residue of the next chain", you're taking "the last residue of the current chain + 1".
Won't they always be the same?
```
h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain + 1]] ?? nResidues
h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain + 1] - 1] + 1
```

Reply:
Technically, this should be the case. But your solution does out-of-bounds access to an array and extra if check for undefined and null for each assignment.